c10/core/TensorImpl.cpp

```diff
@@ -95,6 +97,11 @@ TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::
     key_set_ = key_set | getAutogradRelatedKeySetFromBackend(k);
   }
 
+  // Inference tensor doesn't have version counter.
+  if (this->is_inference_tensor()) {
+    version_counter_ = VariableVersion(/*enabled=*/false);
```

Comment:
I don't know if the compiler was able to figure this out, but what you have in the code here is that first the default constructor for the `version_counter_` field is run (which allocates a new intrusive pointer), and then you immediately afterwards unset it with a nullptr. I suppose you can check what actually happens in the code here with godbolt, but it seems unlikely to me that the compiler will know how to elide the initial allocation if `is_inference_tensor()` is true.

What you should do instead is default initialize the VariableVersion field to nullptr (the cheap initialize), and then only later when you determine that this is an inference tensor.

Reply:
In fact, I would go a step further. Today, we unconditionally allocate a version counter in the constructor for TensorImpl, but in view situations, we then replace it with the shared version counter from the view base. This is also a waste! We could just directly pass in the intended version counter to the TensorImpl constructor. I'm not sure how much refactoring it would take to do this (maybe should have been done before this patch), but it would then make the situation here a lot clearer.